crates/rethnet_evm/src/db/fork.rs

```diff
@@ -0,0 +1,271 @@
+use hashbrown::HashMap;
+use revm::{db::DatabaseRef, Account, AccountInfo, Bytecode};
+
+use rethnet_eth::{Address, B256, U256};
+
+use crate::db::{
+    layered_db::{LayeredDatabase, RethnetLayer},
+    remote::{RemoteDatabase, RemoteDatabaseError},
+};
+
+/// A database integrating the state from a remote node and the state from a local layered
+/// database.
+pub struct ForkDatabase {
+    layered_db: LayeredDatabase<RethnetLayer>,
+    remote_db: RemoteDatabase,
+    account_info_cache: HashMap<Address, AccountInfo>,
+    code_by_hash_cache: HashMap<B256, Bytecode>,
+    storage_cache: HashMap<(Address, U256), U256>,
+    fork_block_number: u64,
+    fork_block_state_root_cache: Option<B256>,
+}
+
+/// An error emitted by ForkDatabase
+#[derive(thiserror::Error, Debug)]
+pub enum ForkDatabaseError {
+    /// An error from the underlying RemoteDatabase
+    #[error(transparent)]
+    RemoteDatabase(#[from] RemoteDatabaseError),
+
+    /// An error from the underlying LayeredDatabase
+    #[error(transparent)]
+    LayeredDatabase(#[from] <LayeredDatabase<RethnetLayer> as revm::Database>::Error),
+
+    /// Account not found
+    #[error("Could not find account {0}")]
+    AccountNotFound(Address),
+
+    /// Some other error from an underlying dependency
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+impl ForkDatabase {
+    /// instantiate a new ForkDatabase
+    pub fn new(url: &str, fork_block_number: u64) -> Self {
+        let remote_db = RemoteDatabase::new(url);
+
+        let layered_db = LayeredDatabase::<RethnetLayer>::default();
+
+        Self {
+            layered_db,
+            remote_db,
+            account_info_cache: HashMap::new(),
+            code_by_hash_cache: HashMap::new(),
+            storage_cache: HashMap::new(),
+            fork_block_number,
+            fork_block_state_root_cache: None,
+        }
+    }
+}
+
+impl revm::Database for ForkDatabase {
+    type Error = ForkDatabaseError;
+
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        if let Some(cached) = self.account_info_cache.get(&address) {
+            Ok(Some(cached.clone()))
+        } else if let Some(layered) = self
+            .layered_db
+            .basic(address)
+            .map_err(ForkDatabaseError::LayeredDatabase)?
+        {
+            Ok(Some(layered))
+        } else if let Some(remote) = self
+            .layered_db
```

Comment:
I had said there wasn't anything in [the n-api PR](https://github.com/NomicFoundation/hardhat/pull/3612) that touched `ForkDatabase`, but I remembered that there is this one small (but glaring haha) thing. This should be `remote_db`, not `layered_db`. I fixed it over [here](https://github.com/NomicFoundation/hardhat/pull/3612/commits/fc3047457fb1b50b732c34a3a358c1cdae375fe8) but haven't ported it back into this PR yet.

Reply:
Is it cherry pickable?

Reply:
yup, picked in https://github.com/NomicFoundation/hardhat/pull/3575/commits/359d3e67cdae3cba56e9355ddd0cf0e51e7bf7cf